howtos/module/_posts/2016-07-07-Common-mistakes-while-coding-module.md

```diff
@@ -0,0 +1,183 @@
+
+- Always check the input data: use `Tools::getValue('var')` instead of `$_POST['var']`, `$_GET['var']`.
+- [Type hinting](http://php.net/manual/en/language.oop5.typehinting.php) makes sure the value has the correct type: use `(int)`, `(bool)`, `(float)` properly, check the public methods available in the [`classes/Validate.php`](https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Validate.php), such as  [`isUnsignedInt()`](https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Validate.php#L686-L695).
+- The PrestaShop token is your friend: always check the token when doing action for logged in customers. Need an example?
+
+{% gist kpodemski/13a33eba6b43e9b6db7d29e491104ae1 %}
+
+  You can also check the token for guests users: I suggested [these two methods](https://github.com/PrestaShop/PrestaShop/pull/5863/files).
+
+- Secure your files: files outside your module's base file (such as `ajax.php`, `upload.php`, etc.) should have a check at their very beginning in order to make sure they cannot be triggered directly. Here is example:
+
+{% gist kpodemski/5717d7c36d4277a71fa67746bb8efade %}
+
+
+### 4. Setup module compatibility and restrictions
+
+Let's be honest, sometimes it's hard to maintain a module from version to version, and sometimes you'd like to release a module only for specific versions of PrestaShop, even specific countries!
+
+How to set the compatibility of your module? In your `__construct` method:
+
+`$this->ps_versions_compliancy = array('min' => '1.6.1.0', 'max' => _PS_VERSION_);`
+
+This code will not allow to install the module if you have PrestaShop in version lower than 1.6.1.0.
+
+How to restrict your module for specific countries? In your `__construct` method:
+
+`$this->limited_countries = array('fr', 'pl');`
+
+This code will not allow to install the module in countries other than France and Poland. Use [ISO-3166 codes](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2#Officially_assigned_code_elements)!
+
+
+### 5. Do not trigger code if it is for a specific page
+
+If you are using the `displayHeader` hook to add features to the Product Page (for instance), do not use this code if you're not on page where you have a single product.
+
+You can easily contextualize your code by using simple `if` statements:
+
+{% highlight php %}
+public function hookDisplayHeader($params)
+{
+    // A good practice is to stop code as soon as possible if we don't need it.
+    if ($this->context->controller->php_self !== 'product') {
+        return;
+    }
+
+    // Rest of the code
+}
+{% endhighlight %}
+
+Quick tip: it's worth noting that if you want to get some information from the Product object, you can do this with this code:
+`$product = $this->context->controller->getProduct()`.<br/>
+There is no need to do `new Product(...)`. This code will work on a single-product page.
+
+
+### 6. Too much going on in your __constructor
```

Comment:
\_\_ makes the text go bold. Try with `\_\_constructor` (in all instances of the word)?